Add a component for GreenEye Monitor

jkeljo opened the pull request:
## Description:

[GreenEye Monitor](http://www.brultech.com/greeneye/) is an energy
monitor that can monitor emergy usage of individual circuits, count pulses
from things like water or gas meters, and monitor temperatures. This component
exposes these various sensors in Home Assistant, for both data tracking and
automation triggering purposes.

**Related issue (if applicable):** n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#6177

## Example entry for `configuration.yaml` (if applicable):
```yaml
greeneye_monitor:
  port: 8000
  monitors:
    - serial_number: YOUR_SERIAL_NUMBER
      channels:
        - number: 1
          name: total_power
        - number: 2
          name: solar_panels_power
          net_metering: true
      pulse_counters:
        - number: 1
          name: sprinklers_water_usage
          counted_quantity: "gal"
          counted_quantity_per_pulse: 1
          time_unit: "min"
      temperature_sensors:
        - number: 1
          name: back_porch_temperature
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

jkeljo committed:
Add a component for GreenEye Monitor

[GreenEye Monitor](http://www.brultech.com/greeneye/) is an energy
monitor that can monitor emergy usage of individual circuits, count pulses
from things like water or gas meters, and monitor temperatures. This component
exposes these various sensors in Home Assistant, for both data tracking and
automation triggering purposes.

MartinHjelmare reviewed (changes requested):
Consolidate the platforms into one sensor platform that adds the correct entity type based on the component config schema option sent as discovery info.

jkeljo committed:
Consolidate sensors

jkeljo committed:
lint

jkeljo committed:
.coveragerc

jkeljo commented:
Sorry for the delay! Rebased on a more recent dev branch and refactored as requested.

jkeljo committed:
- cv.ensure_list
- DOMAIN, where appropriate
- defaults to schema
- single invocation of async_load_platform
- async_create_task instead of async_add_job
- fail if no sensors
- monitors required
- async_add_entities
- call add_devices once
- remove unused schema
- use properties rather than set fields
- move _number and unique_id to GEMSensor
- remove unnecessary get(xxx, None)
- keep params on one line when possible
- new-style string format

jkeljo commented on homeassistant/components/greeneye_monitor.py:
```diff
@@ -0,0 +1,164 @@
+PULSE_COUNTER_SCHEMA = vol.Schema({
+    vol.Required(CONF_NUMBER): vol.Range(1, 4),
+    vol.Required(CONF_NAME): cv.string,
+    vol.Required(CONF_COUNTED_QUANTITY): cv.string,
```
The pulse counter channels accept pulses from any kind of sensor that emits pulses. There's no limit on what the sensor might be counting -- liters, watt-hours, sheep... :)

MartinHjelmare commented on homeassistant/components/greeneye_monitor.py:
```diff
@@ -0,0 +1,169 @@
+
+    if not all_sensors:
+        _LOGGER.error("Configuration must specify at least one "
+                      "channel, pulse counter or temperature sensor.")
```
We don't end logging messages with period.

MartinHjelmare reviewed (commented):
Looks great. One comment.

MartinHjelmare commented on homeassistant/components/greeneye_monitor.py:
```diff
@@ -0,0 +1,169 @@
+
+TEMPERATURE_SENSORS_SCHEMA = vol.Schema({
+    vol.Required(CONF_TEMPERATURE_UNIT): cv.temperature_unit,
+    vol.Required(CONF_SENSORS): cv.ensure_list(TEMPERATURE_SENSOR_SCHEMA),
```
Does this work both when passing a dict within a list and just a dict?

We normally do:
```py
vol.Required(CONF_SENSORS): vol.All(
    cv.ensure_list(), [TEMPERATURE_SENSOR_SCHEMA]),
```

jkeljo replied:
Good catch. Fixed.

MartinHjelmare reviewed (commented):
One question too.

jkeljo committed:
Fix `ensure_list` usage, log message

jkeljo committed:
Pass config through

MartinHjelmare reviewed (approved):
Great! Can be merged when build passes.

jkeljo commented:
Rebased on `dev` to see if that makes coverage happy

MartinHjelmare reviewed (approved):
:tada:
